### ISSUE TITLE
Several bugs in drag'n'dropping items

### ISSUE THREAD

**Issue by olivierkes** [olivierkes/manuskript#169]
I found several bugs happening when **drag'n'dropping items, or deleting / creating them**. I think the last commits I pushed (a44dd41f008e577f6f9ff5772a9e8a804e5d187e and 648b4b67ebedbd5e5d7a08285b20b253b9e1885c mainly) fix following bugs:

# 1. Critical: crash or loss of data

## a) An item dropped into itself (or children) would be silently deleted

![test](https://user-images.githubusercontent.com/1192933/31850072-c5be8ad2-b64c-11e7-83f1-529c20f77c86.gif)

**Reproduce (simple):**

- Open a folder
- Drag it (from the tree view) onto itself

**Reproduce (more complex):**

- Split view
- Open a folder `A` into one view (folder `A` must be in another folder)
- Select the other split view
- Move one of the parent's folder of folder `A` onto `A`

**Result:** the folder is deleted 

**After fix:** you're not allowed to drop

## b) Sometimes after deleting a item, manuskript would crash when creating a new item

**Steps to reproduce:**

- Create a folder
- Create a text
- Move text into folder
- In tree view, select text (it opens in editor)
- In tree view, select folder
- Press the `remove item` button (at the bottom of tree view), item is deleted
- Press the `new folder` or `new text` button (at the bottom of tree view) → manuskript crashed

**After fix:** well, it doesn't crash :)

## c) Random crash when drag'n'dropping wildly

Sometime manuskript used to crash randomly. I have been moving things around and not observed any crashes yet.

# 2. Annoying UI issues

## a) When moving a text item, the tab name is changed to root

![test](https://user-images.githubusercontent.com/1192933/31850101-5271749e-b64d-11e7-9d1a-61916093b78f.gif)

This is more than simply annoying: it meant that a view could cache two different references (one `QModelIndex` and one `outlineItem`'s ID) towards two different items. It had the potential to create bugs.

**After fix:** tab name is right (but item is not selected in treeView, this has still to be fixed)

## b) Sometimes when creating / moving / deleting an item, the folder view would switch to Root

**Steps to reproduce:**

- Select a text file within a folder
- Delete text by clicking on the `remove item` button in tree view
- The view swtiches to root

This happened also randomely when creating a item sometimes.

**After fix:** views switch to parent folder (but again, is not selected in treeView)

# Conclusion

If you want to help, please:

- Confirm the above fixes work for you (in the **future** branch), and that I have not broken other things
- Test recklessly manuskript for thoses kind of moves, and see if you spot a) crashes or b) unintuitive/unclear behavior

Thanks :)

**Comment by gedakc** [comment]
@olivierkes do these drag 'n drop enhancements apply to the **Plots** pane?

Using the _feature/future_ branch with last commit f09c961399, I ran the tests from issue #134 and there are still problems with drag 'n drop with plot steps.

**Comment by olivierkes** [comment]
Right I wasn't clear. I tested only the moments in **Editor** for text/folders items. Not plots. Still have to do that.

**Comment by gedakc** [comment]
While testing the _feature/future_ branch with last commit 761286c497 I discovered a critical bug.  When dragging a folder and dropping it between two same folder split panes the folder is silently deleted.

For example in the screen shot below, drag **Chapter 8** between the split panes on the right.  Drop the folder between the panes when the icon changes from a circle with a cross to a hand icon.  The folder will be silently deleted.

![Manuskript Editor Split Screen Drag Bug](https://gparted.org/curtis/manuskript/Manuskript-Editor-Split-Screen-Drag-Bug.png)

**Comment by olivierkes** [comment]
Well spotted @gedakc. I don't know why this happens, but I found a way to prevent it (I think). Thanks !

**Comment by gedakc** [comment]
Good work @olivierkes on fixing the silent delete when a folder was dragged and dropped between two same folder split panes.

I have confirmed that this issue is fixed with the _feature/future_ branch with last commit f99e4facab.

**Comment by olivierkes** [comment]
Should we include those commits in 0.5.0? 

And while we are at it, the whole future branch? It adds following features : cursor blinking, custom editor margins, custom icons, tree view icon size. Those seem pretty safe to me. It also corrects a few minor bugs.

**Comment by gedakc** [comment]
> Should we include those commits in 0.5.0?

Yes, that sounds good to me.  The enhancements that prevent silent deletion of folders are especially important.

**Comment by gedakc** [comment]
I am testing with the _feature/future_ branch with last commit f99e4facab and I just found another bug with drag 'n drop.  The problem is that if a folder containing text scenes is open during the drag operation, the text scenes will be duplicated on the drop operation.

For example shift-select all the chapters and scenes as shown:

![Manuskript Editor - Before DragNDrop](https://gparted.org/curtis/manuskript/Manuskript-Editor-Before-DragNDrop.png)

Now drag these and drop above the **Book** folder.

Notice that the scenes in **Chapter 1** are duplicated above the Chapter 1 folder.

![Manuskript Editor - After DragNDrop](https://gparted.org/curtis/manuskript/Manuskript-Editor-After-DragNDrop.png)

**Comment by olivierkes** [comment]
> If it is okay with you then I would like to test any fix you have for this problem first before including the feature/future branch in 0.5.0. I have other commitments this weekend so it will be Monday before I get a chance to test again. Hopefully you will have some time for testing as well.

@gedakc  Sure, we are in no hurry. Better to take a bit more time, do more test, and create a better release. I'll do more testing too.

**Comment by olivierkes** [comment]
Found another critical bug, when **moving around folders with the same name** while saving to multiple files (not single file). For example, dragging from:
![image](https://user-images.githubusercontent.com/1192933/32089335-a9d905b2-bae9-11e7-93d0-396966a76316.png)
To:
![image](https://user-images.githubusercontent.com/1192933/32089438-4abf0f80-baea-11e7-9e95-21ba431148dc.png)

We get an `OSError` and a crash:

```
Traceback (most recent call last):
  File "/home/olivier/Dropbox/Documents/Travail/Geekeries/Python/PyCharmProjects/manuskript/bin/../manuskript/mainWindow.py", line 591, in saveDatas
    r = loadSave.saveProject()  # version=0
  File "/home/olivier/Dropbox/Documents/Travail/Geekeries/Python/PyCharmProjects/manuskript/bin/../manuskript/loadSave.py", line 21, in saveProject
    return v1.saveProject()
  File "/home/olivier/Dropbox/Documents/Travail/Geekeries/Python/PyCharmProjects/manuskript/bin/../manuskript/load_save/version_1.py", line 353, in saveProject
    os.replace(oldPath, newPath)
OSError: [Errno 39] Directory not empty: '/home/olivier/Dropbox/Documents/Travail/Geekeries/Python/PyCharmProjects/manuskript/test-projects/toolbar/outline/1-Folder' -> '/home/olivier/Dropbox/Documents/Travail/Geekeries/Python/PyCharmProjects/manuskript/test-projects/toolbar/outline/0-Folder'
Fatal Python error: Aborted

Current thread 0x00002aaaaaaf2400 (most recent call first):
  File "/home/olivier/Dropbox/Documents/Travail/Geekeries/Python/PyCharmProjects/manuskript/bin/../manuskript/main.py", line 76 in launch
  File "/home/olivier/Dropbox/Documents/Travail/Geekeries/Python/PyCharmProjects/manuskript/bin/../manuskript/main.py", line 64 in run
  File "bin/manuskript", line 13 in <module>
make: *** [run] Aborted (core dumped)
```

So that's a bug in the saving algorithm.

**Comment by olivierkes** [comment]
https://github.com/olivierkes/manuskript/commit/276559bc049bda2df5d1944d2a799a814f5cfa7c should fix previous bug, *except in one case*.

The exception is in the following scenario:

1. With manuskript prior to https://github.com/olivierkes/manuskript/commit/276559bc049bda2df5d1944d2a799a814f5cfa7c, save a project containing items with identical names (as in screenshot above)
2. In manuskript https://github.com/olivierkes/manuskript/commit/276559bc049bda2df5d1944d2a799a814f5cfa7c, open said project, and *before saving* (therefore before doing anything else, if autosave is enabled) change the order of items and save. 

In other terms, it means that manuskript would **not** crash if

- Items with identical names are created with manuskript https://github.com/olivierkes/manuskript/commit/276559bc049bda2df5d1944d2a799a814f5cfa7c
- Items with identical names are created with manuskript prior https://github.com/olivierkes/manuskript/commit/276559bc049bda2df5d1944d2a799a814f5cfa7c but the project is saved at least once before moving items around.

I could probably prevent this bug, but that would mean introducing code only for that very specific and quite improbable case, so I have half a mind to leave it at that.

**Comment by gedakc** [comment]
Thank you @olivierkes for your continued improvements to drag 'n drop.  I have confirmed that the issues I had encountered earlier are now resolved.

I think these improvements would be great additions to the upcoming 0.5.0 release.

Regarding whether to add time specific code, I leave that decision up to you.  I think the most important thing is that our users do not lose data.  For example if there is some way for users to prevent a potential crash by editing their data, then that would be worthwhile posting in the release notes.

**Comment by olivierkes** [comment]
Thanks for testing!

I think we should merge the whole feature/future into develop before releasing. Do you agree?

**Comment by gedakc** [comment]
> I think we should merge the whole feature/future into develop before releasing. Do you agree?

If you wish to merge all of the changes into develop before releasing 0.5.0, then I am okay with that.

My main concern is that some enhancements might include changes to the base English language strings.  I don't think it is fair to the translators to alter these base strings _after_ we asked them to update the language translations.

**Comment by olivierkes** [comment]
> My main concern is that some enhancements might include changes or additions to the base English language strings. I don't think it is fair to the translators to alter these base strings after we asked them to update the language translations.

Yes that was my concern too. But I think I'd rather have manuskript 0.5.0 with the feature/future improvements and a few untranslated strings (probably < 1%), than a fully translated manuskript and those commits waiting till next release.

So I take it on me to merge, while saying once more my deep gratitude to translators ! :)

**Comment by gedakc** [comment]
I will be standing by ready to build the Linux and Windows PyInstaller packages.  :-)

**Comment by TheHeadlessSourceMan** [comment]
Just got a crash while dragging a folder containing multiple items  within the editor view.

release version 0.7.0 win7 64bit

Backtrace:
`Traceback (most recent call last):
  File "manuskript\models\abstractModel.py", line 125, in parent
  File "manuskript\models\abstractItem.py", line 124, in row
ValueError: <manuskript.models.outlineItem.outlineItem object at 0x0FE5D410> is not in list
`

After restarting the program, attempting to drag the same folder to the same place worked fine. :/
Not exactly reproducible, but perhaps that backtrace will help at least.

**Comment by gedakc** [comment]
Thank you for your interest in improving Manuskript.  We are aware of some issues with copying/moving.  The specific issues mentioned in this report were addressed, included in 0.5.0, and this issue was closed.  

If you have more information then please look at currently open issues, or create a new issue with details on how to reproduce the problem.
